rules/ruleset.go

```diff
@@ -683,3 +685,524 @@ type int64Asc []int64
+		}
+	}
+	rs.updateMetadata(meta)
+	return nil
+}
+
+func (rs *ruleSet) UpdateRollupRule(rru RollupRuleUpdate) error {
+	rrd := rru.Data
+	err := rs.validateRollupRuleUpdate(rrd)
+	if err != nil {
+		return err
+	}
+
+	r, err := rs.getRollupRuleByID(rru.ID)
+	if err != nil {
+		return fmt.Errorf(ruleActionErrorFmt, "update", rru.ID, err)
+	}
+
+	meta := rrd.UpdateMetadata
+	if err = r.addSnapshot(
+		rrd.Name,
+		rrd.Filters,
+		rrd.Targets,
+		meta.cutoverNanos,
+		rs.tagsFilterOpts,
+	); err != nil {
+		return fmt.Errorf(ruleActionErrorFmt, "update", rrd.Name, err)
+	}
+	rs.updateMetadata(meta)
+	return nil
+}
+
+func (rs *ruleSet) DeleteRollupRule(d DeleteData) error {
+	r, err := rs.getRollupRuleByID(d.ID)
+	if err != nil {
+		return fmt.Errorf(ruleActionErrorFmt, "delete", d.ID, err)
+	}
+	meta := d.UpdateMetadata
+	if err := r.markTombstoned(meta.cutoverNanos); err != nil {
+		return fmt.Errorf(ruleActionErrorFmt, "delete", d.ID, err)
+	}
+	return nil
+}
+
+func (rs *ruleSet) Delete(meta UpdateMetadata) error {
+	if rs.tombstoned {
+		return fmt.Errorf("%s is already tombstoned", string(rs.namespace))
+	}
+
+	rs.tombstoned = true
+	rs.updateMetadata(meta)
+
+	// Make sure that all of the rules in the ruleset are tombstoned as well.
+	for _, m := range rs.mappingRules {
+		if t := m.Tombstoned(); !t {
+			m.markTombstoned(meta.cutoverNanos)
+		}
+	}
+
+	for _, r := range rs.rollupRules {
+		if t := r.Tombstoned(); !t {
+			r.markTombstoned(meta.cutoverNanos)
+		}
+	}
+
+	return nil
+}
+
+func (rs *ruleSet) Revive(meta UpdateMetadata) error {
+	if !rs.Tombstoned() {
+		return fmt.Errorf(ruleSetActionErrorFmt, "revive", string(rs.namespace), errNotTombstoned)
+	}
+
+	rs.tombstoned = false
+	rs.updateMetadata(meta)
+	return nil
+}
+
+func (rs *ruleSet) updateMetadata(meta UpdateMetadata) {
+	rs.cutoverNanos = meta.cutoverNanos
+	rs.lastUpdatedAtNanos = meta.lastUpdatedAtNanos
+}
+
+func (rs ruleSet) getMappingRuleByName(name string) (*mappingRule, error) {
+	for _, m := range rs.mappingRules {
+		n, err := m.Name()
+		if err != nil {
+			continue
+		}
+
+		if n == name {
+			return m, nil
+		}
+	}
+	return nil, errNoSuchRule
+}
+
+func (rs ruleSet) getMappingRuleByID(id string) (*mappingRule, error) {
+	for _, m := range rs.mappingRules {
+		if m.uuid == id {
+			return m, nil
+		}
+	}
+	return nil, errNoSuchRule
+}
+
+func (rs ruleSet) getRollupRuleByName(name string) (*rollupRule, error) {
+	for _, r := range rs.rollupRules {
+		n, err := r.Name()
+		if err != nil {
+			return nil, err
+		}
+
+		if n == name {
+			return r, nil
+		}
+	}
+	return nil, errNoSuchRule
+}
+
+func (rs ruleSet) getRollupRuleByID(id string) (*rollupRule, error) {
+	for _, r := range rs.rollupRules {
+		if r.uuid == id {
+			return r, nil
+		}
+	}
+	return nil, errNoSuchRule
+}
+
+//RuleConflictError is returned when a rule modification is made that would conflict with the current state.
+type RuleConflictError struct {
+	ConflictUUID string
```

Comment:
I'd like to be explicit about what that uuid is. In this case its the conflicting id not the id of the rule you are trying to add. 

This is exported because it will be used later in m3ctl to return different error codes/messages when there is a conflict as opposed to some other type of error.